### PR TITLE
EES-2137 - fixed remaining failing UI tests after front-end changes

### PR DESCRIPTION
**Pull request by duncan-at-hiveit**
This PR:
- fixes currently-failing UI tests
- adds arguments to run_tests.py to rerun failed tests and suites and combine results with original run
- adds arguments to run_tests.py to specify own implicit and explicit wait times for Robot runs
- adds the generation of an HTML dump to accompany screenshots when failures occur.  From these it is easier to find out the cause of failures if down to difference in HTML structure.

An example usage of running with the extra options:

`pipenv run python3 run_tests.py -i robot -b chrome -e local -v --rerun-failed-suites --timeout 10 --implicit-wait 2` 

In this particular example, this is re-running only the suites that failed in the previous run, and then merging any new successes with the original run results and creating a new report.html.

### CODE REVIEW

File: src/explore-education-statistics-frontend/README.md [hunk 1]
```diff
@@ -55,6 +55,8 @@ Your app is ready to be deployed!
 
 See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.
 
+Locally you can also run `npm run start:prod` to run the production-ready bundle.
```
Comment (luke-h1) [inline src/explore-education-statistics-frontend/README.md:58]: It'd be best to make this stand out more and provide some more info on how the team typically uses this command (aside from deployment). So you could do something such as: 
``` 
### `npm run start:prod`
Builds the app for production. This command can also be helpful if you are experiencing flakiness when running tests locally

File: tests/robot-tests/run_tests.py [hunk 2]
```diff
@@ -90,6 +91,22 @@
                     dest="disable_teardown",
                     help="disable tearing down of any test data after completion",
                     action='store_true')
+parser.add_argument("--rerun-failed-tests",
+                    dest="rerun_failed_tests",
+                    action='store_true',
+                    help="rerun individual failed tests and merge results into original run results")
+parser.add_argument("--rerun-failed-suites",
```
Comment (luke-h1) [inline tests/robot-tests/run_tests.py:98]: Loving these changes! 🚀

File: tests/robot-tests/tests/admin_and_public/bau/prerelease_visibility.robot [hunk 3]
```diff
@@ -54,7 +54,8 @@ Go to 'Sign Off' page
     user clicks link  Sign off
     user waits for page to finish loading
     user waits until page contains testid  public-release-url
-    ${PUBLIC_RELEASE_LINK}=  Get Text  xpath://*[@data-testid="public-release-url"]
+    ${PUBLIC_RELEASE_LINK}=  Get Value  xpath://*[@data-testid="public-release-url"]
+    check that variable is not empty  PUBLIC_RELEASE_LINK   ${PUBLIC_RELEASE_LINK}
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public/bau/prerelease_visibility.robot:58]: Nice! That'll be a handy keyword for the future

File: tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot [hunk 4]
```diff
@@ -721,7 +719,7 @@ Generate table
 
 Wait until the table is generated
     [Tags]  HappyPath
-    user waits until page contains button  Generate permanent link
+    user clicks button  Share your table
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:722]: You'll probably want to what until the button is rendered (dev can sometimes be slow depending on the amount of people using it). So you'll want to add the below line above `user clicks button Share your table` 

`user waits until page contains button. Share your table  60`

File: tests/robot-tests/tests/libs/admin-common.robot [hunk 5]
```diff
@@ -29,7 +29,7 @@ user signs in as analyst1
 
     user waits until h1 is visible  Dashboard
     user waits until page contains title caption  Welcome Analyst1
-    user waits until page contains element   id:publicationsReleases-themeTopic-themeId   180
+    user waits until page contains element   css:#publicationsReleases-themeTopic-themeId,[data-testid='no-permission-to-access-releases']   180
```
Comment (luke-h1) [inline tests/robot-tests/tests/libs/admin-common.robot:32]: Don't think the comma here is necessary?

**Review by luke-h1 — CHANGES_REQUESTED** [review]

File: tests/robot-tests/tests/libs/admin-common.robot (hunk 5 above)
Comment (markhiveit) [reply tests/robot-tests/tests/libs/admin-common.robot:32]: I think this is necessary. The selector is either looking for either `#publicationsReleases-themeTopic-themeId` or `[data-testid='no-permission-to-access-releases']`. The first appears if analyst1 does have permission view/edit releases - the second appears is analyst1 has no permission to view/edit any releases. This covers both cases.

File: tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot (hunk 4 above)
Comment (markhiveit) [reply tests/robot-tests/tests/admin_and_public/bau/publish_release_and_amend_2.robot:722]: `user clicks button` contains `user gets button element` which in turn calls `user waits until parent contains element` - so an additional wait shouldn't be necessary.

File: tests/robot-tests/tests/libs/admin-common.robot (hunk 5 above)
Comment (luke-h1) [reply tests/robot-tests/tests/libs/admin-common.robot:32]: ah that's pretty cool. Didn't realise that was possible

**Comment by luke-h1** [comment]
UI test changes look good. However, it looks like some unit tests are failing. I think the snapshots will need updating

**Review by markhiveit — APPROVED** [review]

**Pull request merged**